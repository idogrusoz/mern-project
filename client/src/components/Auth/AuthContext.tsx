import React, {
    createContext,
    FunctionComponent,
    Dispatch,
    useMemo,
    useState,
    SetStateAction,
    ReactNode,
    useCallback,
} from "react";
import api from "../../api";
import { ResponseUser } from "../../../../interfaces/user.interfaces";

export type AuthContextObject = {
    isAuthenticated: () => Promise<void>;
    authenticated: boolean;
    setAuthenticated: Dispatch<SetStateAction<boolean>>;
    user: ResponseUser | null;
    setUser: Dispatch<SetStateAction<ResponseUser | null>>;
    signOut: () => Promise<void>;
};

type Props = {
    children: ReactNode;
};

export const defaultAuthContextValue: AuthContextObject = {
    isAuthenticated: () => new Promise(() => false),
    authenticated: false,
    setAuthenticated: () => {},
    user: null,
    setUser: () => {},
    signOut: () => new Promise(() => {}),
};

const AuthContext = createContext<AuthContextObject>(defaultAuthContextValue);

const AuthContextProvider: FunctionComponent<Props> = ({ children }) => {
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [user, setUser] = useState<ResponseUser | null>(null);

    const isAuthenticated = useCallback(async (): Promise<void> => {
        try {
            const response = await api.get("auth/authenticate");
            setAuthenticated(response.data.isAuthenticated);
            if (!user) {
                setUser(response.data.user);
            }
        } catch (error) {
            setAuthenticated(false);
            setUser(null);
        }
    }, [user]);

    const signOut = useCallback(async () => {
        try {
            await api.get("auth/signout");
            setAuthenticated(false);
            setUser(null);
        } catch (err) {
            console.log("Error : ", err);
            throw err;
        }
    }, []);

    const AuthContextValue: AuthContextObject = useMemo(() => {
        const value = {
            isAuthenticated,
            authenticated,
            setAuthenticated,
            user,
            setUser,
            signOut,
        };
        return value;
    }, [isAuthenticated, authenticated, setAuthenticated, user, setUser, signOut]);

    return <AuthContext.Provider value={AuthContextValue}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, AuthContext };
