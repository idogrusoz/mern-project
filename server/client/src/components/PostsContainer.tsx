import { useMediaQuery } from "@material-ui/core";
import React, { FunctionComponent } from "react";
import { IPostInterface } from "../../../interfaces/post.interfaces";
import PostDisplay from "./PostDisplay";

type PostContainerProps = {
    posts: IPostInterface[];
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({ posts }): JSX.Element => {
    const isBigScreen = useMediaQuery("(min-width: 650px)");
    return (
        <div
            style={
                !isBigScreen
                    ? { height: "calc(100vh - 170px)", overflowY: "auto", overflowX: "visible" }
                    : { height: "calc(100vh - 70px)", overflowY: "auto", overflowX: "visible" }
            }
        >
            {posts.map((post: IPostInterface, i: number) => {
                return <PostDisplay post={post} key={i} />;
            })}
        </div>
    );
};

export default PostsContainer;
