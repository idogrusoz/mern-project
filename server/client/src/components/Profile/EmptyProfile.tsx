import { Button, Grid, makeStyles, Typography } from "@material-ui/core";
import React, { FunctionComponent, useContext } from "react";
import { ProfileContext } from "./ProfileContext";

const useStyles = makeStyles({
    item: { textAlign: "center", marginTop: "30%" },
});

const EmptyProfile: FunctionComponent<{}> = () => {
    const { profileOwner } = useContext(ProfileContext);
    const classes = useStyles();
    return (
        <>
            <Grid container justify="center" alignContent="center">
                <Grid item xs={12}>
                    <Typography variant="h3" className={classes.item}>
                        {profileOwner ? profileOwner.displayName : "You"} don't have any posts yet
                    </Typography>
                </Grid>
                {!profileOwner && (
                    <Grid item xs={12} className={classes.item}>
                        <Button variant="contained" color="primary">
                            Add Post
                        </Button>
                    </Grid>
                )}
            </Grid>
        </>
    );
};

export default EmptyProfile;
