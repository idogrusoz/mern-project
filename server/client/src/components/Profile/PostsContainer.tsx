import React, { FunctionComponent } from "react";
import { BasePostDocument } from "../../../../interfaces/post.interfaces";
import PostDisplay from "../PostDisplay";

type PostContainerProps = {
    posts: BasePostDocument[];
    fetch: () => Promise<void>;
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({ posts, fetch }): JSX.Element => {
    return (
        <div>
            {posts.map((post: BasePostDocument, i: number) => {
                return <PostDisplay post={post} fetch={fetch} />;
            })}
        </div>
    );
};

export default PostsContainer;
