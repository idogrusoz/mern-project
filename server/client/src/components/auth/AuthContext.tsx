import React, {
    createContext,
    FunctionComponent,
    Dispatch,
    useMemo,
    useState,
    SetStateAction,
    ReactNode,
    useCallback,
} from "react";
import api from "../../api";
import { ResponseUser } from "../../../../interfaces/user.interfaces";

type AuthContextObject = {
    // isAuthenticated: () => Promise<void>;
    authenticated: boolean;
    setAuthenticated: Dispatch<SetStateAction<boolean>>;
    user: ResponseUser | null;
    setUser: Dispatch<SetStateAction<ResponseUser | null>>;
    logOut: () => Promise<void>;
};

type Props = {
    children: ReactNode;
};

const defaultCotextValue: AuthContextObject = {
    // isAuthenticated: () => new Promise(() => false),
    authenticated: false,
    setAuthenticated: () => {},
    user: null,
    setUser: () => {},
    logOut: () => new Promise(() => {}),
};

const AuthContext = createContext<AuthContextObject>(defaultCotextValue);

const AuthContextProvider: FunctionComponent<Props> = ({ children }) => {
    const [authenticated, setAuthenticated] = useState<boolean>(false);
    const [user, setUser] = useState<ResponseUser | null>(null);

    // Tobe uncommented once API route s built
    // const isAuthenticated = useCallback(async (): Promise<void> => {
    //     try {
    //         const response = await api.get("auth/authenticated");
    //         console.log(response.status);
    //         setAuthenticated(response.data.isAuthenticated);
    //         if (!user) {
    //             setUser(response.data.user);
    //         }
    //     } catch (error) {
    //         console.log(error);
    //         setAuthenticated(false);
    //         setUser(null);
    //     }
    // }, [user]);

    // useEffect(() => {
    //     const fetch = () => {
    //         isAuthenticated();
    //     };
    //     fetch();
    // }, [isAuthenticated]);

    const logOut = useCallback(async () => {
        try {
            await api.get("auth/logout");
        } catch (err) {
            console.log("Error : ", err);
            throw err;
        }
    }, []);

    const AuthContextValue: AuthContextObject = useMemo(() => {
        const value = {
            // isAuthenticated,
            authenticated,
            setAuthenticated,
            user,
            setUser,
            logOut,
        };
        return value;
    }, [
        // isAuthenticated,
        authenticated,
        setAuthenticated,
        user,
        setUser,
        logOut,
    ]);

    return <AuthContext.Provider value={AuthContextValue}>{children}</AuthContext.Provider>;
};

export { AuthContextProvider, AuthContext };
