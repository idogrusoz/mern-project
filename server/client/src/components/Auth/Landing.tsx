import { Container, Typography } from "@material-ui/core";
import React, { useCallback, useContext, useState } from "react";
import { SignInUser, UserInterface } from "../../../../interfaces/user.interfaces";
import api from "../../api";
import { AuthContext } from "./AuthContext";
import Register from "./Register";
import SignIn from "./SignIn";

const Landing = () => {
    const [showSignin, setShowSignin] = useState<boolean>(true);
    const [error, setError] = useState<boolean>(false);
    const [errorMessage, setErrorMessage] = useState<string>("");
    const { setAuthenticated, setUser } = useContext(AuthContext);

    const register = useCallback(async (user: UserInterface) => {
        try {
            const response = await api.post("auth/register", user);
            setShowSignin(response.status === 200);
        } catch (error) {
            setError(true);
            setErrorMessage(error.message);
        }
    }, []);

    const signin = useCallback(
        async (user: SignInUser) => {
            try {
                const response = await api.post("auth/signin", user);
                setAuthenticated(response.data.isAuthenticated);
                setUser(response.data.user);
            } catch (error) {
                setError(true);
                setErrorMessage(error.message);
            }
        },
        [setAuthenticated, setUser],
    );

    return (
        <Container maxWidth="sm">
            {error && <Typography color="secondary">{errorMessage}</Typography>}
            {showSignin ? (
                <SignIn signin={signin} setShowSignin={setShowSignin} />
            ) : (
                <Register register={register} setShowSignin={setShowSignin} />
            )}
        </Container>
    );
};

export default Landing;
