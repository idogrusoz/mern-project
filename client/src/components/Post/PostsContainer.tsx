import { CircularProgress } from "@material-ui/core";
import React, { FunctionComponent, useEffect } from "react";
import { BasePostDocument } from "../../../../interfaces/post.interfaces";
import PostDisplay from "./PostDisplay";
import EmptyProfile from "../Profile/EmptyProfile";

export type PostContainerProps = {
    posts: BasePostDocument[];
    fetchLikedPosts: () => Promise<void>;
    fetchFeed: () => Promise<void>;
    fetchUserPosts: () => Promise<void>;
    activeTab: "feed" | "posts" | "likes";
    isLoading: boolean;
};

const PostsContainer: FunctionComponent<PostContainerProps> = ({
    posts,
    fetchLikedPosts,
    fetchFeed,
    fetchUserPosts,
    activeTab,
    isLoading,
}): JSX.Element => {
    useEffect(() => {
        if (activeTab === "feed") {
            fetchFeed();
        }
        if (activeTab === "posts") {
            fetchUserPosts();
        }
        if (activeTab === "likes") {
            fetchLikedPosts();
        }
    }, [activeTab, fetchFeed, fetchUserPosts, fetchLikedPosts]);
    return isLoading ? (
        <div style={{ display: "flex", justifyContent: "center" }}>
            <CircularProgress />
        </div>
    ) : (
        <>
            {posts.length > 0 ? (
                posts.map((post: BasePostDocument, i: number) => {
                    return <PostDisplay post={post} fetch={fetchLikedPosts} key={i} />;
                })
            ) : (
                <EmptyProfile activeTab={activeTab} />
            )}
        </>
    );
};

export default PostsContainer;
