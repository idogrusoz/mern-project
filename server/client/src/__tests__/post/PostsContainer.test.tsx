import React from "react";
import toJson from "enzyme-to-json";
import { mount, render, shallow } from "enzyme";
import PostsContainer, { PostContainerProps } from "../../components/Post/PostsContainer";
import { MemoryRouter } from "react-router-dom";
import { AuthContext, AuthContextObject, defaultContextValue } from "../../components/Auth/AuthContext";
import EmptyProfile from "../../components/Profile/EmptyProfile";
const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useHistory: () => ({
        push: mockHistoryPush,
    }),
}));
describe("PostsContainer test", () => {
    const props: PostContainerProps = {
        activeTab: "feed",
        fetchFeed: jest.fn(),
        fetchLikedPosts: jest.fn(),
        fetchUserPosts: jest.fn(),
        isLoading: false,
        posts: [
            {
                _id: "postId",
                likes: [],
                textContent: "",
                author: {
                    user_id: "userId",
                    image: "",
                    userName: "userName",
                },
                style: {
                    fontFamily: "Roboto",
                    color: "#ffffff",
                    backgroundColor: "#ffffff",
                    fontWeight: 400,
                    fontSize: 40,
                    textAlign: "left" as "left" | "right" | "justify" | "center",
                },
            },
        ],
    };

    const authValue: AuthContextObject = {
        ...defaultContextValue,
        user: {
            _id: "userId",
            firstName: "firstName",
            lastName: "lastName",
            displayName: "displayName",
            userName: "userName",
            email: "email",
        },
    };

    const mountComponent = (props: PostContainerProps) => {
        return mount(
            <AuthContext.Provider value={authValue}>
                <MemoryRouter>
                    <PostsContainer {...props} />
                </MemoryRouter>
            </AuthContext.Provider>,
        );
    };

    it("renders without crash", () => {
        const wrapper = shallow(<PostsContainer {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it("calls feed function to fetch posts", () => {
        const tabs: Array<"feed" | "posts" | "likes"> = ["feed", "posts", "likes"];
        for (let tab of tabs) {
            const newProps: PostContainerProps = { ...props, activeTab: tab };
            mountComponent(newProps);
            if (tab === "feed") {
                expect(props.fetchFeed).toHaveBeenCalled();
            }
            if (tab === "posts") {
                expect(props.fetchUserPosts).toHaveBeenCalled();
            }
            if (tab === "likes") {
                expect(props.fetchLikedPosts).toHaveBeenCalled();
            }
        }
    });
    it("displays empty profile component when there are no posts", () => {
        const propsWithNoPosts: PostContainerProps = { ...props, posts: [] };
        const wrapper = mountComponent(propsWithNoPosts);
        expect(wrapper.find(EmptyProfile).length).toBe(1);
    });
});
