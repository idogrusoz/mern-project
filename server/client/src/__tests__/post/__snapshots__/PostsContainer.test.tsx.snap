// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsContainer test renders without crash 1`] = `
<Fragment>
  <PostDisplay
    fetch={[MockFunction]}
    key="0"
    post={
      Object {
        "_id": "postId",
        "author": Object {
          "image": "",
          "userName": "userName",
          "user_id": "userId",
        },
        "likes": Array [],
        "style": Object {
          "backgroundColor": "#ffffff",
          "color": "#ffffff",
          "fontFamily": "Roboto",
          "fontSize": 40,
          "fontWeight": 400,
          "textAlign": "left",
        },
        "textContent": "",
      }
    }
  />
</Fragment>
`;
